sqldb/assets.go

```diff
@@ -0,0 +1,393 @@
+		if err != nil {
+			return nil, err
+		}
+		assets = listRowsToAssets(aus)
+	}
+	var firstID uint = 0
+	var lastID uint = 0
+	if len(assets) > 0 {
+		firstID = uint(assets[0].ID)
+		lastID = uint(assets[len(assets)-1].ID)
+	}
+
+	dl := domain.ListedAssets{
+		FirstID: firstID,
+		LastID:  lastID,
+		Limit:   query.Limit,
+		Assets:  assets,
+	}
+	return &dl, nil
+}
+
+func (d *DB) FavouriteAsset(ctx context.Context, userID, assetID uint, at domain.AssetType, isFavourite bool) (uint, error) {
+	var nid uint = 0
+	var err error
+	switch at {
+	case domain.InsightAssetType:
+		if isFavourite {
+			count := int64(0)
+			d.db.Model(FavouriteInsight{}).Where("user_id = ? AND insight_id = ? ", userID, assetID).Count(&count)
+			if count == 0 {
+				in := &FavouriteInsight{UserID: userID, InsightID: assetID}
+				err = d.db.Create(in).Error
+				nid = in.ID
+			} else {
+				err = errors.New("record exists")
+			}
+		} else {
+			in := &FavouriteInsight{}
+			err = d.db.Where("user_id = ? AND insight_id = ? ", userID, assetID).Unscoped().Delete(in).Error
+		}
+	case domain.ChartAssetType:
+		if isFavourite {
+			count := int64(0)
+			d.db.Model(FavouriteChart{}).Where("user_id = ? AND chart_id = ? ", userID, assetID).Count(&count)
+			if count == 0 {
+				ch := &FavouriteChart{UserID: userID, ChartID: assetID}
+				err = d.db.Create(ch).Error
+				nid = ch.ID
+			} else {
+				err = errors.New("record exists")
+			}
+		} else {
+			in := &FavouriteChart{}
+			err = d.db.Where("user_id = ? AND chart_id = ? ", userID, assetID).Unscoped().Delete(in).Error
+		}
+	case domain.AudienceAssetType:
+		if isFavourite {
+			count := int64(0)
+			d.db.Model(FavouriteAudience{}).Where("user_id = ? AND audience_id = ? ", userID, assetID).Count(&count)
+			if count == 0 {
+				au := &FavouriteAudience{UserID: userID, AudienceID: assetID}
+				err = d.db.Create(au).Error
+				nid = au.ID
+			} else {
+				err = errors.New("record exists")
+			}
+		} else {
+			in := &FavouriteAudience{}
+			err = d.db.Where("user_id = ? AND audience_id = ? ", userID, assetID).Unscoped().Delete(in).Error
+		}
+	}
+	if err != nil {
+		return 0, err
+	}
+
+	return nid, nil
+}
+
+func (d *DB) listFavouriteAudiences(ctx context.Context, userID uint, onlyFav bool, query domain.QueryAssets) (*domain.ListedAssets, error) {
+	var aus []AudienceWithFavour
+	gormQuery := d.db.Model(Audience{}).Select("audiences.*, (favourite_audiences.user_id = ?) AS is_favourite", userID)
+	if onlyFav {
+		if query.IsDesc {
+			gormQuery = gormQuery.Joins("INNER JOIN favourite_audiences ON favourite_audiences.audience_id = audiences.id AND audiences.id < ? AND favourite_audiences.user_id = ?", query.LastID, userID).Order("audiences.id desc")
+		} else {
+			gormQuery = gormQuery.Joins("INNER JOIN favourite_audiences ON favourite_audiences.audience_id = audiences.id AND audiences.id > ? AND favourite_audiences.user_id = ?", query.LastID, userID).Order("audiences.id asc")
```

Comment:
These 2 parts of the condition differ only in the comparison operator `< >` - how would you approach abstracting the query?

Reply:
When I started writing the SQL code, I was wondering if they were going to become multiline so I can put them in .tpl files and use the template package to add the values.
But I endured the long strings without thinking too much how to improve them because of the personal time restriction at the time of the development.
However, for this specific code I would rewrite it to something like this
```go
ineq := ">"
order := "asc"
if isDesc {
   ineq = "<"
   order = "desc"
}

gormQuery = gormQuery.Joins(fmt.Sprintf("INNER JOIN favourite_audiences ON favourite_audiences.audience_id = audiences.id AND audiences.id %s ? AND favourite_audiences.user_id = ?",ineq ), query.LastID, userID).Order("audiences.id "+order)
```